lib/collection/src/shards/shard_holder.rs

```diff
@@ -37,44 +39,78 @@ pub const SHARD_KEY_MAPPING_FILE: &str = "shard_key_mapping.json";
+///
+/// Uses a `SmallVec` putting two IDs on the stack. That's the maximum number of shards we expect
+/// with the current resharding implementation.
+pub type ShardIds = SmallVec<[ShardId; 2]>;
+
 pub struct ShardHolder {
     shards: HashMap<ShardId, ShardReplicaSet>,
     pub(crate) shard_transfers: SaveOnDisk<HashSet<ShardTransfer>>,
-    rings: HashMap<RingsKey, HashRing<ShardId>>,
+    rings: HashMap<Option<ShardKey>, ShardHashRing>,
     key_mapping: SaveOnDisk<ShardKeyMapping>,
     // Duplicates the information from `key_mapping` for faster access
     // Do not require locking
     shard_id_to_key_mapping: HashMap<ShardId, ShardKey>,
 }
 
-pub type LockedShardHolder = RwLock<ShardHolder>;
-
-#[derive(Clone, Debug, Eq, PartialEq, Hash)]
-pub enum RingsKey {
-    Default,
-    ShardKey(ShardKey),
-    Resharding,
+pub enum ShardHashRing {
+    /// Single hashring
+    Single(HashRing<ShardId>),
+    /// Two hashrings when transitioning during resharding
+    /// Depending on the current resharding state, points may be in either or both shards.
+    Resharding {
+        old: HashRing<ShardId>,
+        new: HashRing<ShardId>,
+    },
 }
 
-impl From<Option<ShardKey>> for RingsKey {
-    fn from(shard_key: Option<ShardKey>) -> Self {
-        match shard_key {
-            Some(shard_key) => shard_key.into(),
-            None => Self::Default,
+impl ShardHashRing {
+    pub fn is_empty(&self) -> bool {
+        match self {
+            Self::Single(ring) => ring.is_empty(),
+            Self::Resharding { old, new } => old.is_empty() && new.is_empty(),
+        }
+    }
+
+    pub fn add(&mut self, shard: ShardId) {
+        match self {
+            Self::Single(ring) => ring.add(shard),
+            // When resharding, update the new ring
+            Self::Resharding { old: _, new } => new.add(shard),
+        }
+    }
+
+    pub fn get<U: Hash>(&self, key: &U) -> ShardIds {
+        match self {
+            Self::Single(ring) => ring.get(key).into_iter().cloned().collect(),
+            Self::Resharding { old, new } => old
+                .get(key)
+                .into_iter()
+                .chain(new.get(key))
+                // Both hash rings may return the same shard ID, take it once
+                .dedup()
+                .cloned()
+                .collect(),
         }
     }
 }
 
-impl From<ShardKey> for RingsKey {
-    fn from(shard_key: ShardKey) -> Self {
-        Self::ShardKey(shard_key)
+impl From<HashRing<ShardId>> for ShardHashRing {
+    fn from(ring: HashRing<ShardId>) -> Self {
+        Self::Single(ring)
     }
 }
 
+pub type LockedShardHolder = RwLock<ShardHolder>;
+
 impl ShardHolder {
     pub fn new(collection_path: &Path) -> CollectionResult<Self> {
-        let mut rings = HashMap::new();
-        rings.insert(RingsKey::Default, HashRing::fair(HASH_RING_SHARD_SCALE));
+        let rings = HashMap::from([(
+            None,
+            ShardHashRing::from(HashRing::fair(HASH_RING_SHARD_SCALE)),
```

Comment:
I'd add `ShardHashRing::new` or `ShardHashRing::single` method for this. It's used in a bunch of places, seems annoying to have to type this every time.

Reply:
Sounds good.

Reply:
Implemented in ddb2066ee41c4b09a6dfdaa0d316c2793c4d9cb8